jvm/src/main/java/com/microsoft/bond/protocol/CompactBinaryProtocolWriter.java

```diff
@@ -0,0 +1,383 @@
+package com.microsoft.bond.protocol;
+
+import com.microsoft.bond.BondDataType;
+import com.microsoft.bond.Metadata;
+import com.microsoft.bond.ProtocolType;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.Queue;
+import java.util.Stack;
+
+// TODO: add Compact Binary protocol format comment (either here and/or in the deserializer and/or in package.html)
+
+/**
+ * Implements Compact Binary protocol (serialization).
+ */
+public final class CompactBinaryProtocolWriter implements TwoPassProtocolWriter {
+
+    private static final short MAGIC = (short)ProtocolType.COMPACT_PROTOCOL.value;
+
+    private final BinaryStreamWriter writer;
+    private final short version;
+    private final FirstPassCounter firstPassCounter;
+
+    private int currentStructIndex;
+
+    public CompactBinaryProtocolWriter(OutputStream outputStream, short version) {
+        this.writer = new BinaryStreamWriter(outputStream);
+        this.version = version;
+        this.firstPassCounter = (version == 2) ? new FirstPassCounter() : null;
```

Comment:
As discussed with the broader team, we'll fix that elsewhere in code. I am adding check for valid version.

---
In reply to: [116131706](https://github.com/tstein/bond/pull/4#discussion_r116131706) [](ancestors = 116131706,116114875)